include/LightGBM/utils/ChunkedArray.hpp

```diff
@@ -0,0 +1,260 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ *
+ * Author: Alberto Ferreira
+ */
+#ifndef __CHUNKED_ARRAY_H__
+#define __CHUNKED_ARRAY_H__
+
+#include <stdint.h>
+#include <assert.h>
+
+#include <new>
+#include <vector>
+#include <algorithm>
+
+
+/**
+ * Container that manages a dynamic array of fixed-length chunks.
+ *
+ * The class also takes care of allocation & release of the underlying
+ * memory. It can be used with either a high or low-level API.
+ *
+ * The high-level API allocates chunks as needed, manages addresses automatically and keeps
+ * track of number of inserted elements, but is not thread-safe (this is ok as usually input is a streaming iterator).
+ * For parallel input sources the low-level API must be used.
+ *
+ * Note: When using this for `LGBM_DatasetCreateFromMats` use a
+ *       chunk_size multiple of #num_cols for your dataset, so each chunk
+ *       contains "complete" instances.
+ *
+ * === High-level insert API intro ===
+ *
+ * The easiest way to use is:
+ *  0. ChunkedArray(chunk_size)  # Choose appropriate size
+ *  1. add(value)                # as many times as you want (will generate chunks as needed)
+ *  2. data() or void_data()     # retrieves a T** or void** pointer (useful for `LGBM_DatasetCreateFromMats`).
+ *
+ * Useful query methods (all O(1)):
+ *  - get_add_count()   # total count of added elements
+ *  - get_chunks_count()  # how many chunks are currently allocated.
+ *  - get_current_chunk_added_count()  # for the last add() chunk, how many items there are.
+ *  - get_chunk_size()    # Get constant chunk_size from constructor call.
+ *
+ *  With those you can generate int32_t sizes[]. Last chunk can be smaller than chunk_size, so, for any i:
+ *    - sizes[i<last] = get_chunk_size()
+ *    - sizes[i==last] = get_add_count()
+ *
+ *
+ * === Low-level insert API intro ===
+ *
+ * For advanced usage - useful for inserting in parallel - one can also:
+ *  1. call new_chunk() at any time for as many chunks as needed.  (thread-UNsafe)
+ *  2. call setitem(chunk, idx, value) to insert each value.       (thread-safe)
+ *
+ */
+template <class T>
+class ChunkedArray {
+ public:
+    explicit ChunkedArray(size_t chunk_size)
+      : _chunk_size(chunk_size), _last_chunk_idx(0), _last_idx_in_last_chunk(0) {
+       new_chunk();
+    }
+
+    ~ChunkedArray() {
+        release();
+    }
+
+    /**
+     * Adds a value to the chunks sequentially.
+     * If the last chunk is full it creates a new one and appends to it.
+     *
+     * @param value value to insert.
+     */
+    void add(T value) {
+        if (!within_bounds(_last_chunk_idx, _last_idx_in_last_chunk)) {
+            new_chunk();
+            ++_last_chunk_idx;
+            _last_idx_in_last_chunk = 0;
+        }
+
+        assert(setitem(_last_chunk_idx, _last_idx_in_last_chunk, value) == 0);
```

Comment:
@StrikerRUS  I did `#include <LightGBM/utils/log.h>` but got this:

![image](https://user-images.githubusercontent.com/3976753/111142317-73ca2a80-857c-11eb-9ce2-3407f5c24974.png)

which is defined in that header file.

Any idea how to fix this?

Reply:
Sorry, I'm not a cpp coder, but maybe the reason in the namespace?
https://github.com/microsoft/LightGBM/blob/296b2a2662a1560c5baa5a0e6ec32cbe290654b9/include/LightGBM/utils/log.h#L25

Reply:
I agreed. Adding `namespace LightGBM` should allow use to use `CHECK_EQ` directly. And in `test_ChunkedArray.cpp`, we may need
```cpp
template <typename T>
using ChunkedArray = LightGBM::ChunkedArray<T>;
```
to avoid tedious `LightGBM::ChunkedArray`.

By the way, can we change the file names `ChunkedArray.hpp` and `test_ChunkedArray.cpp` into something like `chunked_array.hpp` to keep them consistent with other file names in the code base? Just a minor suggestion.

Thanks for your hard work! @AlbertoEAF

Reply:
Didn't see the namespace, thanks!